packages/lumx-react/src/components/tabs/TabProvider.tsx

```diff
@@ -0,0 +1,154 @@
+import get from 'lodash/get';
+import set from 'lodash/set';
+import startCase from 'lodash/startCase';
+import React, { ReactNode, createContext, useContext, useEffect, useState } from 'react';
+import uuid from 'uuid/v4';
+
+export interface TabProviderProps {
+    /** Active tab index. */
+    activeTabIndex?: number;
+    /** Tab provider children. */
+    children: ReactNode;
+    /** Tab panel children should not render if the tab panel is hidden. */
+    isLazy?: boolean;
+    /** Activate tabs on focus. */
+    shouldActivateOnFocus?: boolean;
+
+    /** Tab change callback. */
+    onChange?(index: number): void;
+}
+
+const DEFAULT_PROPS: Partial<TabProviderProps> = {
+    activeTabIndex: 0,
+    isLazy: true,
+    shouldActivateOnFocus: false,
+};
+
+interface TabState {
+    index: number;
+    isActive: boolean;
+    tabId?: string;
+    tabPanelId?: string;
+}
+
+type SetTabState = (tabState: TabState) => void;
+
+/** State stored the tab context. */
+export interface TabContextState extends Omit<Required<TabProviderProps>, 'children' | 'activeTabIndex'> {
+    /**
+     * Hook to be called by the Tab component.
+     * This hook will register the tab on mount and unregister on unmount.
+     * If the tab does not provide an id, a uuid base id will be generated.
+     *
+     * @return the tab id and tab index.
+     */
+    useTab(id?: string): TabState;
+
+    /**
+     * Hook to be called by the TabPanel component.
+     * This hook will register the tab panel on mount and unregister on unmount.
+     * If the tab panel does not provide an id, a uuid base id will be generated.
+     *
+     * @return the tab panel id and tab panel index.
+     */
+    useTabPanel(id?: string): TabState;
+}
+
+const TabContext = createContext<TabContextState | null>(null);
+type TabProvider = React.FC<TabProviderProps> & { useContext(): TabContextState };
+
+/**
+ * This component provides a context in which tabs can be defined and linked to their tab panel.
+ *
+ * It does not produce any markup so you can wrap it around any React elements and then split the TabList and TabPanel
+ * components in the react tree.
+ *
+ * @param  props React component props.
+ * @return React element.
+ */
+export const TabProvider: TabProvider = (props) => {
+    const { children, shouldActivateOnFocus, activeTabIndex, isLazy, onChange } = props;
+    const [state] = useState<TabContextState>(
```

Comment:
Having that much logic in a useState seems really odd to me...
I might be wrong, but couldn't we have used a useReducer instead ?

Reply:
I probably would have started with a useReducer if I knew what I was getting into with this component ^^
But I am indeed thinking how to adapt into a reducer. It's a shame we don't have slice in this repo :/